File: app/models/talk.rb
# frozen_string_literal: true

class Talk < ApplicationRecord
  belongs_to :room

  validates :title, presence: true, length: {maximum: 50}
  validates :title, uniqueness: {scope: :room_id, message: I18n.t('errors.messages.taken')}
  validates :talker_name, presence: true, length: {maximum: 50}

  scope :ordered_by_number, -> { order('order_number desc NULLS LAST, id asc') }

  enum progress: {entried: 0, started: 1, oneFifths: 2, twoFifths: 3, threeFifths: 4, fourFifths: 5, done: 6}

  def json_attributes(room)
    attributes.merge(room: room.attributes)
  end
end

File: app/javascript/rooms/models/talk.js
import Immutable from 'immutable';

const TalkBase = Immutable.Record({
  entries: new Immutable.List(),
  done: new Immutable.List(),
  loading: true,
  current: new Immutable.Map(),
});

export const progress = {
  entried: { value: 0, key: 'entried' },
  oneFifths: { value: 1, key: 'oneFifths' },
  twoFifths: { value: 2, key: 'twoFifths' },
  threeFifths: { value: 3, key: 'threeFifths' },
  fourFifths: { value: 4, key: 'fourFifths' },
  done: { value: 5, key: 'done' },
};

export const progressPercent = {
  entried: 0,
  oneFifths: 20,
  twoFifths: 40,
  threeFifths: 60,
  fourFifths: 80,
  done: 100,
};

export default class Talk extends TalkBase {
}
